File: poserp/lib/modules/products/services/product_service.dart
import '../../../core/api/api_client.dart';
import '../../../core/api/api_endpoints.dart';
import '../../../data/models/api_response.dart';
import '../models/product.dart';
import '../models/product_payload.dart';

class ProductService {
  final ApiClient _apiClient;

  ProductService(this._apiClient);

  Future<ApiResponse<List<Product>>> getAll({
    String? search,
    String? category,
    int page = 1,
    int limit = 15,
  }) async {
    final response = await _apiClient.get(
      ApiEndpoints.products,
      queryParameters: {
        if (search != null && search.isNotEmpty) 'search': search,
        if (category != null && category.isNotEmpty && category != 'all')
          'category': category,
        'page': page,
        'limit': limit,
      },
    );

    return ApiResponse<List<Product>>.fromJson(
      response.data,
      (json) => (json as List).map((item) => Product.fromJson(item)).toList(),
    );
  }

  Future<ApiResponse<Product>> getById(String id) async {
    final response = await _apiClient.get('${ApiEndpoints.products}/$id');
    return ApiResponse<Product>.fromJson(
      response.data,
      (json) => Product.fromJson(json),
    );
  }

  Future<ApiResponse<Product>> create(ProductPayload payload) async {
    final response = await _apiClient.post(
      ApiEndpoints.products,
      data: payload.toJson(),
    );

    return ApiResponse<Product>.fromJson(
      response.data,
      (json) => Product.fromJson(json),
    );
  }

  Future<ApiResponse<Product>> update(String id, ProductPayload payload) async {
    final response = await _apiClient.put(
      '${ApiEndpoints.products}/$id',
      data: payload.toJson(),
    );

    return ApiResponse<Product>.fromJson(
      response.data,
      (json) => Product.fromJson(json),
    );
  }

  Future<void> delete(String id) async {
    await _apiClient.delete('${ApiEndpoints.products}/$id');
  }
}

File: src/services/productService.ts
import api from "./api";
import type { Product, ApiResponse, Pagination, ProductPriceOptionsResponse } from "@/types";
import { db } from "@/lib/db";

export const productService = {
  getAll: async (params?: {
    page?: number;
    limit?: number;
    search?: string;
    category?: string;
    sortBy?: string;
    sortOrder?: string;
    lowStock?: boolean;
  }): Promise<{ data: Product[]; pagination: Pagination }> => {
    try {
      const { data } = await api.get<ApiResponse<Product[]>>("/products", { params });
      
      // Cache to IndexedDB asynchronously
      if (data.data && data.data.length > 0 && !params?.search && !params?.category) {
        // Only cache the full list (or first big chunk) when there are no filters
        db.products.bulkPut(data.data).catch(console.error);
      }
      
      return { data: data.data, pagination: data.pagination! };
    } catch (error: any) {
      // If offline, fetch from Dexie
      if (error.code === 'ERR_NETWORK' || !navigator.onLine) {
        console.log("Offline mode: Fetching products from local DB");
        let query = db.products.toCollection();
        
        if (params?.search) {
          const searchLower = params.search.toLowerCase();
          query = db.products.filter(p => 
            p.name.toLowerCase().includes(searchLower) || 
            (p.barcode || "").toLowerCase().includes(searchLower) || 
            (p.sku || "").toLowerCase().includes(searchLower)
          );
        } else if (params?.category) {
          query = db.products.filter(p => p.category === params.category);
        }
        
        const localProducts = await query.toArray();
        return { 
          data: localProducts, 
          pagination: { 
            total: localProducts.length, 
            page: 1, 
            limit: localProducts.length, 
            pages: 1 
          } 
        };
      }
      throw error;
    }
  },

  getById: async (id: string): Promise<Product> => {
    const { data } = await api.get<ApiResponse<Product>>(`/products/${id}`);
    return data.data;
  },

  getByBarcode: async (barcode: string): Promise<Product> => {
    try {
      const { data } = await api.get<ApiResponse<Product>>(`/products/barcode/${barcode}`);
      if (data.data) {
        db.products.put(data.data).catch(console.error);
      }
      return data.data;
    } catch (error: any) {
      if (error.code === 'ERR_NETWORK' || !navigator.onLine) {
        const localProduct = await db.products.where('barcode').equals(barcode).first();
        if (localProduct) return localProduct;
      }
      throw error;
    }
  },

  getPriceOptions: async (id: string): Promise<ProductPriceOptionsResponse> => {
    const { data } = await api.get<ApiResponse<ProductPriceOptionsResponse> & ProductPriceOptionsResponse>(`/products/${id}/price-options`);
    return data.data || {
      product: data.product,
      priceOptions: data.priceOptions,
      defaultPrice: data.defaultPrice,
      priceSelectionRequired: data.priceSelectionRequired,
    };
  },

  create: async (payload: Partial<Product>): Promise<Product> => {
    const { data } = await api.post<ApiResponse<Product>>("/products", payload);
    return data.data;
  },

  update: async (id: string, payload: Partial<Product>): Promise<Product> => {
    const { data } = await api.put<ApiResponse<Product>>(`/products/${id}`, payload);
    return data.data;
  },

  delete: async (id: string): Promise<void> => {
    await api.delete(`/products/${id}`);
  },

  getStats: async () => {
    const { data } = await api.get("/products/stats/overview");
    return data.data;
  },

  getPricing: async (id: string, strategy: "latest" | "fifo" = "latest"): Promise<{
    productId: string;
    batchNo: string;
    purchasePrice: number;
    salesPrice: number;
    availableQty: number;
    taxPercent: number;
    salesTaxType?: string;
  }> => {
    const { data } = await api.get<ApiResponse<any>>(`/products/${id}/pricing`, { params: { strategy } });
    return data.data;
  },

  bulkImport: async (products: any[]): Promise<any> => {
    const { data } = await api.post<ApiResponse<any>>("/products/bulk-import", { products });
    return data;
  },

  getGlobalLibrary: async (params?: { search?: string; barcode?: string }): Promise<any[]> => {
    const { data } = await api.get<ApiResponse<any[]>>("/products/global-library", { params });
    return data.data;
  },
};
